Fail on invalid and unknown ROS specific arguments

hidmic opened the pull request:
Precisely what the title says. This pull request makes `rclpy` raise if either an invalid or an unknown ROS argument is found.

hidmic commented:
We should probably discuss the implications of having exceptions defined in the C extensions vs. Python sources.

hidmic commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -317,9 +320,70 @@ _rclpy_parse_args(PyObject * pyargs, rcl_arguments_t * parsed_args)
+
+    pyargs = PySequence_List(pyargs);
+    if (pyargs == NULL) {
+      goto inner_cleanup;
```
Yeah, I understand what you're saying but it doesn't allow me to keep variable definitions close to the place of first use. So I refactored it entirely. PTAL.

hidmic commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -317,9 +320,70 @@ _rclpy_parse_args(PyObject * pyargs, rcl_arguments_t * parsed_args)
+      }
+      if (PyList_Append(unknown_ros_pyargs, ros_pyarg) != 0) {
+        goto inner_cleanup;
+      }
```
You're right, but that's because I'm not using the right API. I should be doing `PyList_GetItem` (which borrows a reference) instead of `PySequence_GetItem` (which creates a new reference).

hidmic commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -317,9 +320,70 @@ _rclpy_parse_args(PyObject * pyargs, rcl_arguments_t * parsed_args)
+    }
+
+    for (int i = 0; i < unparsed_ros_args_count; ++i) {
+      PyObject * ros_pyarg = PySequence_GetItem(pyargs, (Py_ssize_t)i);
```
Oops, good catch!

hidmic commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -317,9 +320,70 @@ _rclpy_parse_args(PyObject * pyargs, rcl_arguments_t * parsed_args)
+      }
+    }
+
+    PyErr_Format(
```
AFAIK consensus is to fail on unknown ROS arguments at the `rclpy`/`rclcpp` level. Both client libraries are consistent in this regard.

hidmic commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -4975,5 +5039,34 @@ static struct PyModuleDef _rclpymodule = {
+    Py_DECREF(m);
+    return NULL;
+  }
+  PyModule_AddObject(m, "RCLInvalidROSArgsError", RCLInvalidROSArgsError);
```
Good point.

hidmic commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -4975,5 +5039,34 @@ static struct PyModuleDef _rclpymodule = {
+  }
+  PyModule_AddObject(m, "UnknownROSArgsError", UnknownROSArgsError);
+
+  if (PyErr_Occurred()) {
```
I'll do both. It's never too much error checking :)

ivanpauno commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -289,6 +289,48 @@ _rclpy_pyargs_to_list(PyObject * pyargs, int * num_args, char *** arg_values)
+
+  pyargs = PySequence_List(pyargs);
+  if (NULL == pyargs) {
+    goto cleanup;
```
Before going to cleanup, set a generic `RuntimeError` in all the cases.

hidmic replied:
AFAIK all these functions set an error (to be raised) on failure.

ivanpauno replied:
Yes, you're right!

ivanpauno commented:
@hidmic This have introduced a linter failure (from #420)
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=8088)](http://ci.ros2.org/job/ci_linux/8088/)

 Where is the original CI build of this branch?

hidmic commented:
@ivanpauno it's [here](https://github.com/ros2/rcl/pull/493#issuecomment-527564932). Only reasonable explanation is that I thought I re-ran it after addressing your comments but I didn't. I'll fix it, sorry!
